src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs

```diff
@@ -0,0 +1,452 @@
+                return default(string);
+            }
+            if (end._block == _block)
+            {
+                return _utf8.GetString(_block.Array, _index, end._index - _index);
+            }
+
+            var decoder = _utf8.GetDecoder();
+
+            var length = GetLength(end);
+            var charLength = length * 2;
+            var chars = new char[charLength];
+            var charIndex = 0;
+
+            var block = _block;
+            var index = _index;
+            var remaining = length;
+            while (true)
+            {
+                int bytesUsed;
+                int charsUsed;
+                bool completed;
+                var following = block.End - index;
+                if (remaining <= following)
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        remaining,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        true,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    return new string(chars, 0, charIndex + charsUsed);
+                }
+                else if (block.Next == null)
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        following,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        true,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    return new string(chars, 0, charIndex + charsUsed);
+                }
+                else
+                {
+                    decoder.Convert(
+                        block.Array,
+                        index,
+                        following,
+                        chars,
+                        charIndex,
+                        charLength - charIndex,
+                        false,
+                        out bytesUsed,
+                        out charsUsed,
+                        out completed);
+                    charIndex += charsUsed;
+                    remaining -= following;
+                    block = block.Next;
+                    index = block.Start;
+                }
+            }
+        }
+
+        public ArraySegment<byte> GetArraySegment(MemoryPoolIterator2 end)
+        {
+            if (IsDefault || end.IsDefault)
+            {
+                return default(ArraySegment<byte>);
+            }
+            if (end._block == _block)
+            {
+                return new ArraySegment<byte>(_block.Array, _index, end._index - _index);
+            }
+
+            var length = GetLength(end);
+            var array = new byte[length];
```

Comment:
old memory pool probably not going to survive for long... this is the degenerate case also - request headers are expected to very, very rarely span across multiple blocks